### ISSUE TITLE
[Research] Add GNN-Based Spatial-Temporal Graph for 4-Channel EEG Emotion (ASTAGL: 94% cross-subject)

### ISSUE THREAD

**Issue by LakshmiSravyaVedantham** [LakshmiSravyaVedantham/DreamAnalysis#51]
## Paper

**Title:** Adaptive Spatial-Temporal Aware Graph Learning for EEG-Based Emotion Recognition
**Authors:** Various (Cyborg and Bionic Systems / Science Partner Journal)
**Year:** 2025
**Venue:** Cyborg and Bionic Systems (Science Partner Journal)
**URL:** https://spj.science.org/doi/10.34133/cbsystems.0088

**Supporting:** Graph Neural Networks in EEG-based Emotion Recognition: A Survey (arXiv:2402.01138, 2024)

## Key Contribution

ASTAGL learns adaptive spatial graphs that model inter-channel relationships without requiring a fixed electrode layout. Unlike traditional GCNs that use predefined adjacency matrices (requiring 32+ channels), ASTAGL learns the graph structure from data — making it viable for few-channel setups. Achieves 94.22% subject-independent accuracy on SEED (3-class), far exceeding the project's current 74.21% LGBM.

The survey paper (arXiv:2402.01138) documents that GNN approaches consistently outperform CNN/RNN for EEG emotion by 5-15% because they explicitly model spatial dependencies between electrode sites.

## How It Applies to NeuralDreamWorkshop

The current system uses a flat feature vector (band powers + ratios + DASM/RASM) fed to LightGBM. This discards the spatial relationships between TP9-AF7-AF8-TP10. A 4-node graph with learned edge weights can capture:
- **Frontal asymmetry** (AF7 <-> AF8 edge) — currently hand-coded as FAA
- **Temporal-frontal connectivity** (TP9 <-> AF7, TP10 <-> AF8)
- **Cross-hemisphere connections** (TP9 <-> AF8, TP10 <-> AF7)

With only 4 nodes, the graph is tiny (6 possible edges) — inference would be sub-millisecond.

### Files to Modify/Create
- `ml/models/graph_emotion_classifier.py` — new GNN model
- `ml/models/emotion_classifier.py` — add GNN to the fallback chain (above LGBM if accuracy > 74%)
- `ml/training/train_graph_emotion.py` — training script

### Architecture
```python
import torch
import torch.nn as nn
from torch_geometric.nn import GATConv, global_mean_pool

class EEGGraphNet(nn.Module):
    """4-node spatial-temporal graph for Muse 2 EEG emotion."""
    def __init__(self, n_bands=5, n_temporal=4, hidden=64, n_classes=6):
        super().__init__()
        # Node features: DE per band (5) + Hjorth (3) + entropy (1) = 9 per node
        self.node_feat_dim = 9
        # Learnable adjacency (4x4, symmetric)
        self.adj = nn.Parameter(torch.randn(4, 4))
        # Temporal conv per node (captures time dynamics)
        self.temporal_conv = nn.Conv1d(self.node_feat_dim, hidden, kernel_size=3, padding=1)
        # Graph attention layers
        self.gat1 = GATConv(hidden, hidden, heads=4, concat=False)
        self.gat2 = GATConv(hidden, hidden, heads=4, concat=False)
        # Classifier
        self.fc = nn.Sequential(
            nn.Linear(hidden, 32),
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(32, n_classes)
        )

    def forward(self, x, edge_index):
        # x: (batch, 4_nodes, 9_features, T_epochs)
        # Apply temporal conv per node, then GAT across nodes
        ...
```

### Graph Construction for Muse 2
```python
# Static edges (anatomical adjacency)
edges = [
    (0, 1),  # TP9 - AF7 (left hemisphere)
    (2, 3),  # AF8 - TP10 (right hemisphere)
    (1, 2),  # AF7 - AF8 (frontal pair — FAA)
    (0, 3),  # TP9 - TP10 (temporal pair)
]
# + learnable edge weights that adapt during training
```

### Training
- Train on DEAP (32ch, subsample to 4 closest to Muse positions: T7≈TP9, AF3≈AF7, AF4≈AF8, T8≈TP10)
- Fine-tune on any collected Muse data
- Use DE features per channel per band as node features

## Expected Performance Impact

- Current: 74.21% (LGBM, cross-subject, 6-class)
- Expected: 78-85% (cross-subject, 6-class) — conservative estimate for 4-channel
- The survey shows GNNs add +5-15% over flat-feature baselines
- Real benefit: automatically learns FAA/DASM/RASM-equivalent features rather than hand-coding them

## Complexity: MEDIUM

- Requires PyTorch Geometric dependency (already have PyTorch)
- 4-node graph is trivially small — no scalability concerns
- Training: ~1 hour on CPU for DEAP-subsampled data
- Inference: <1ms per epoch (4 nodes, 2 GAT layers)
- Main risk: 4 channels may not provide enough spatial diversity for the graph structure to help

## Dependencies
- `torch-geometric` (pip installable)
- Existing `extract_features_multichannel()` for node feature extraction
